models/base/scratch/snowflake/snowplow_ecommerce_base_events_this_run.sql

```diff
@@ -0,0 +1,77 @@
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_page_1[0]:locale::varchar as ecommerce_page_locale,
+
+  -- unpacking the ecommerce transaction object
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:transaction_id::varchar as transaction_id,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:currency::varchar as transaction_currency,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:payment_method::varchar as transaction_payment_method,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:revenue::decimal(7,2) as transaction_revenue,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:total_quantity::int as transaction_total_quantity,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:credit_order::boolean as transaction_credit_order,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:discount_amount::decimal(7,2) as transaction_discount_amount,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:discount_code::varchar as transaction_discount_code,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:shipping::decimal(7,2) as transaction_shipping,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_transaction_1[0]:tax::decimal(7,2) as transaction_tax,
+
+  -- unpacking the ecommerce cart object
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_cart_1[0]:cart_id::varchar as cart_id,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_cart_1[0]:currency::varchar as cart_currency,
+  a.contexts_com_snowplowanalytics_snowplow_ecommerce_cart_1[0]:total_value::decimal(7,2) as cart_total_value,
+
+  -- unpacking the ecommerce action object
+  a.unstruct_event_com_snowplowanalytics_snowplow_ecommerce_snowplow_ecommerce_action_1:type::varchar as ecommerce_action_type,
+  a.unstruct_event_com_snowplowanalytics_snowplow_ecommerce_snowplow_ecommerce_action_1:name::varchar as ecommerce_action_name,
+
+  a.* exclude(contexts_com_snowplowanalytics_snowplow_web_page_1, domain_userid),
+  dense_rank() over (partition by a.domain_sessionid order by a.derived_tstamp) AS page_view_in_session_index
+
+from {{ var('snowplow__events') }} as a
+inner join {{ ref('snowplow_ecommerce_base_sessions_this_run') }} as b
+on a.domain_sessionid = b.session_id
+
+where a.collector_tstamp <= {{ snowplow_utils.timestamp_add('day', var("snowplow__max_session_days", 3), 'b.start_tstamp') }}
+and a.dvce_sent_tstamp <= {{ snowplow_utils.timestamp_add('day', var("snowplow__days_late_allowed", 3), 'a.dvce_created_tstamp') }}
+and a.collector_tstamp >= {{ lower_limit }}
+and a.collector_tstamp <= {{ upper_limit }}
+and {{ snowplow_utils.app_id_filter(var("snowplow__app_id",[])) }}
+and {{ snowplow_ecommerce.event_name_filter(var("snowplow__ecommerce_event_names", "['snowplow_ecommerce_action']")) }}
+
+qualify row_number() over (partition by a.event_id order by a.collector_tstamp) = 1
```

Comment:
2 things on this (first is all warehouses, second snowflake only). 
1) is it really page view in session_index at this point? Isn't this just event in session index? It feels wrong to call it that if it's not true.
2) For snowflake, my understanding is that window functions take place _before_ qualify does, which would mean you could possibly get different results across BQ and snowflake via this de-duplication. For web it's fine as we don't add any fields, but here we are adding this one and it would differ across warehouses.

Reply:
Great points for both, thanks for catching that Ryan!